feat: state_class (#30)

Kolbi opened the pull request:
Introducte state_class for Sensor entities
- Possibility to define state_class for Sensors
- Currently only the Sensor mileage is defined with state_class="total_increasing",

Kolbi commented:
#333

coreywillwhat commented on custom_components/audiconnect/sensor.py:
```diff
@@ -50,3 +50,8 @@ def device_class(self):
         ):
             return self._instrument.device_class
         return None
+
+    @property
+    def state_class(self):
+        """Return the state_class"""
+        return self._instrument.state_class
```
Should we validate the state_class similar to how `device_class` is done just above this by importing STATE_CLASSES from the same?
https://github.com/home-assistant/core/blob/dev/homeassistant/components/sensor/const.py

of note: looks like DEVICE_CLASSES is deprecated. I will put together a PR to fix that.

Kolbi replied:
Good hint yes we should do it that way.

coreywillwhat commented:
I've tested your code in combination with the validator discussed and it works well. I can add the validator to this if you want, but dont want to overstep either.

It would also be good to add `device_class` for these so we can work on those where needed. I can do that in a separate PR, or you can add to this one.

Let me know.

Kolbi commented:
Please feel free to contribute. You may even edit my PR if you like. I plan being offline from Pc during weekend.

coreywillwhat reviewed (approved):
- added individual `device_class` for all `Sensors`
- added `state_class` validation

Kolbi commented:
Maybe device_class distance could be useful for some sensors. 

Need to check at Pc if others would also make sense:
https://developers.home-assistant.io/docs/core/entity/sensor/

But Prio low from my side currently.

coreywillwhat commented:
I agree it will be useful to update some of the sensors with device_class when we have time. Figure if we rock this in now we can easily go in and update values whenever later.

Thanks for this PR!

cdnninja commented on custom_components/audiconnect/dashboard.py:
```diff
@@ -390,178 +393,302 @@ def create_instruments():
+            name="Mileage",
+            icon="mdi:speedometer",
+            unit="km",
+            state_class="total_increasing",
```
I would think we should be importing these up front vs text and comparison. 

For example:

`from homeassistant.components.sensor import (
    SensorDeviceClass,
    SensorEntity,
    SensorEntityDescription,
    SensorStateClass,
)`

followed by: 

`state_class=SensorStateClass.TOTAL_INCREASING`


Examples are shown here: https://github.com/Hyundai-Kia-Connect/kia_uvo/blob/master/custom_components/kia_uvo/sensor.py

coreywillwhat replied:
oh yeah, this is much cleaner. I'll clean it up.

cdnninja commented on custom_components/audiconnect/dashboard.py:
```diff
@@ -390,178 +393,302 @@ def create_instruments():
+            attr="oil_level",
+            name="Oil level",
+            icon="mdi:oil",
+            unit="%",
```
Unit should be imported from HA as well.

coreywillwhat replied:
on it

cdnninja commented on custom_components/audiconnect/dashboard.py:
```diff
@@ -390,178 +393,302 @@ def create_instruments():
+            attr="charging_power",
+            name="Charging power",
+            icon="mdi:flash",
+            unit="kW",
```
Unit here should also be imported.

cdnninja commented on custom_components/audiconnect/dashboard.py:
```diff
@@ -390,178 +393,302 @@ def create_instruments():
             name="hybrid Range",
             icon="mdi:gas-station-outline",
             unit="km",
+            state_class=None,
```
Any reason to add None when it defaults to this?   Feels to me like longer code.

coreywillwhat replied:
I added the defaults after the fact. which caused it to be redundant. I'll clean those up.

cdnninja commented on custom_components/audiconnect/dashboard.py:
```diff
@@ -390,178 +409,227 @@ def create_instruments():
+        BinarySensor(
+            attr="any_window_open",
+            name="Windows",
+            device_class="window",
```
This would be BinarySensorDeviceClass.WINDOW imported from 

`from homeassistant.components.binary_sensor import (
    BinarySensorDeviceClass,
    BinarySensorEntity,
    BinarySensorEntityDescription,
)`

coreywillwhat replied:
I stuck to just cleaning up the sensors @kolbi started with. Ran out of time and the entire file needs cleanup. Larger cleanup effort might be best under another PR.

Let me know

coreywillwhat replied:
binary_sensors are done

cdnninja commented on custom_components/audiconnect/sensor.py:
```diff
@@ -50,3 +50,13 @@ def device_class(self):
+    def state_class(self):
+        """Return the state_class"""
+        if (
+            self._instrument.state_class is not None
```
I don't think this is needed anymore.

coreywillwhat replied:
Was torn on removing it or not since it's now a benign final check. I can remove it tomorrow

coreywillwhat replied:
removed this and the device class check

cdnninja commented:
Awesome thanks for doing this!!

Kolbi commented:
I'm not sure if this PR but with latest source and my patch branch in combination: https://github.com/Kolbi/audi_connect_ha/releases/tag/v2.0.0a7 I don't get any entities anymore....

coreywillwhat commented:
hmm thats odd, I've had them running for morning without issue. Around the time you commented this though, it started failing to login to the Audi Connect servers...

Kolbi commented:
Argh again API issues. Then it must have been bad timing then older version worked again some minutes later.

coreywillwhat commented:
Yeah idk, I rolled into another version and I log in now but still no data from the cloud. Idk if I'm just testing too much or what but the API has been awful this past week for me.

coreywillwhat commented:
> Argh again API issues. Then it must have been bad timing then older version worked again some minutes later. 

I set the audi_api timeout to 30 instead of 10 and it eventually pulled the data. Getting some serious latency of the data request... Thinking maybe we should build in an increase in the timeout with each retry.

cdnninja commented:
If latency why not just increase the timeout overall?  A retry plus longer timeout increases total time and increases call count.

coreywillwhat commented:
> If latency why not just increase the timeout overall?  A retry plus longer timeout increases total time and increases call count. 

Yeah I'm fried, I need a vacation haha was thinking timeout was a sleep. Different.

I think increasing timeout would make sense. This has happened a bunch recently, and I'm seeing similar latency in the Audi app at the same time. But the Audi app succeeds because longer timeout where HA fails.

Kolbi commented:
Updated to latest master - charging complete time entity works all other entities are deactivated. But can only continue on Monday sadly. But I moved to the old identifier with using the master so maybe that could cause the issue on my side currently.

Kolbi commented:
Seemed to be an issue on my side. Deleted whole integration and readded and all is working again. Still curious why in combination with my changes it was not working. But maybe because of the API issues.
